sopel/coretasks.py

```diff
@@ -540,82 +538,72 @@ def _parse_modes(bot, args, clear=False):
-            value = mapping.get(char)
-            if value is not None:
-                if sign == "+":
-                    priv = priv | value
-                else:
-                    priv = priv & ~value
-                channel.privileges[nick] = priv
-            param_idx += 1
+    privileges = mapping.keys()
+    if 'PREFIX' in bot.isupport:
+        privileges = bot.isupport.PREFIX.keys()
+
+    modemessage = ircmodes.ModeMessage(
+        bot.isupport.CHANMODES,
+        privileges=set(privileges),
+    )
+    modeinfo = modemessage.parse_modestring(args[1], tuple(args[2:]))
+
+    # set or update channel's modes
+    modes = {} if clear else copy.deepcopy(channel.modes)
+    for letter, mode, is_added, param in modeinfo.modes:
+        if letter == 'A':
+            if mode not in modes:
+                modes[mode] = set()
+            if is_added:
+                modes[mode].add(param)
+            elif param in modes[mode]:
+                modes[mode].remove(param)
+        elif letter == 'B':
+            if is_added:
+                modes[mode] = param
+            elif mode in modes:
+                modes.pop(mode)
+        elif letter == 'C':
+            if is_added:
+                modes[mode] = param
+            elif mode in modes:
+                modes.pop(mode)
+        elif letter == 'D':
+            if is_added:
+                modes[mode] = True
+            elif mode in modes:
+                modes.pop(mode)
+
+    # atomic change of channel's modes
+    channel.modes = modes
+
+    # update user privileges in channel
+    for privilege, is_added, param in modeinfo.privileges:
+        # User privs modes, always have a param
+        nick = Identifier(param)
+        priv = channel.privileges.get(nick, 0)
+        value = mapping[privilege]
+        if is_added:
+            priv = priv | value
         else:
-            # Might be in a mode block past A/B/C/D, but we don't speak those.
-            # Send a WHO to ensure no user priv modes we're skipping are lost.
-            LOGGER.warning(
-                "Unknown MODE message, sending WHO. Message was: %r",
-                args,
-            )
-            _send_who(bot, channel_name)
-            return
+            priv = priv & ~value
+        channel.privileges[nick] = priv
 
-    if param_idx != len(params):
+    # log ignored modes (modes Sopel doesn't know how to handle)
+    if modeinfo.ignored_modes:
         LOGGER.warning(
-            "Too many arguments received for MODE: args=%r chanmodes=%r",
+            "Unknown MODE message, sending WHO. Message was: %r",
             args,
-            chanmodes,
         )
+        _send_who(bot, channel_name)
 
-    channel.modes = modes
+    # log leftover parameters (too many arguments)
+    if modeinfo.leftover_params:
+        LOGGER.warning(
+            "Too many arguments received for MODE: args=%r chanmodes=%r",
+            args,
+            modemessage.chanmodes,
+        )
```

Comment:
The mode parser should log this one, not coretasks.

Same could apply to the presence of `ignored_modes`, except coretasks needs that anyway to know that it should `_send_who()`.

Reply:
I really don't like having to depends on logging within the parser itself: the result is a stateless object that contains every bit of required information. For me, it's up to whatever code calls that to know what to do with the result, including dealing with errors.